lib/include/iot_linear_containers.h

```diff
@@ -0,0 +1,888 @@
+ * the list tail are checked. Pass `NULL` to search from the beginning of the list.
+ * @param[in] isMatch Function to determine if an element matches. Pass `NULL` to
+ * search using the address `pMatch`, i.e. `element == pMatch`.
+ * @param[in] pMatch If `isMatch` is `NULL`, each element in the list is compared
+ * to this address to find a match. Otherwise, it is passed as the second argument
+ * to `isMatch`.
+ *
+ * @return Pointer to an #IotLink_t representing the matched and removed element;
+ * `NULL` if no match is found. The macro #IotLink_Container may be used to determine
+ * the address of the link's container.
+ */
+/* @[declare_linear_containers_list_double_removefirstmatch] */
+static inline IotLink_t * IotListDouble_RemoveFirstMatch( IotListDouble_t * const pList,
+                                                          const IotLink_t * const pStartPoint,
+                                                          bool ( *isMatch )( const IotLink_t *, void * ),
+                                                          void * pMatch )
+/* @[declare_linear_containers_list_double_removefirstmatch] */
+{
+    IotLink_t * pMatchedElement = IotListDouble_FindFirstMatch( pList,
+                                                                pStartPoint,
+                                                                isMatch,
+                                                                pMatch );
+
+    if( pMatchedElement != NULL )
+    {
+        IotListDouble_Remove( pMatchedElement );
+    }
+
+    return pMatchedElement;
+}
+
+/**
+ * @brief Remove all matching elements from a doubly-linked list.
+ *
+ * @param[in] pList The doubly-linked list to search.
+ * @param[in] isMatch Function to determine if an element matches. Pass `NULL` to
+ * search using the address `pMatch`, i.e. `element == pMatch`.
+ * @param[in] pMatch If `isMatch` is `NULL`, each element in the list is compared
+ * to this address to find a match. Otherwise, it is passed as the second argument
+ * to `isMatch`.
+ * @param[in] freeElement A function to free memory used by each removed list
+ * element. Optional; pass `NULL` to ignore.
+ * @param[in] linkOffset Offset in bytes of a link member in its container, used
+ * to calculate the pointer to pass to `freeElement`. This value should be calculated
+ * with the C `offsetof` macro. This parameter is ignored if `freeElement` is `NULL`
+ * or its value is `0`.
+ */
+/* @[declare_linear_containers_list_double_removeallmatches] */
+static inline void IotListDouble_RemoveAllMatches( IotListDouble_t * const pList,
+                                                   bool ( *isMatch )( const IotLink_t *, void * ),
+                                                   void * pMatch,
+                                                   void ( *freeElement )( void * ),
+                                                   size_t linkOffset )
```

Comment:
See other comment RE: offset. Too many parameters increase the required stack size, and that is also a perf overhead since stack access may be one order of magnitude slower than register at each access.